src/Microsoft.TestPlatform.Common/ExtensionFramework/Utilities/TestExtensions.cs

```diff
@@ -91,6 +91,74 @@ public class TestExtensions
     /// </summary>
     internal bool AreDataCollectorsCached { get; set; }
 
+    /// <summary>
+    /// Merge two extension maps.
+    /// </summary>
+    /// 
+    /// <param name="extensionMap1">First extension map.</param>
+    /// <param name="extensionMap2">Second extension map.</param>
+    /// 
+    /// <returns>
+    /// A map representing the merger between the two input maps.
+    /// </returns>
+    internal static IDictionary<string, ISet<string>> MergeExtensionMaps(
+        IDictionary<string, ISet<string>> extensionMap1,
+        IDictionary<string, ISet<string>> extensionMap2)
+    {
+        var isExtensionMap1Valid = (extensionMap1 != null && extensionMap1.Count > 0);
+        var isExtensionMap2Valid = (extensionMap2 != null && extensionMap2.Count > 0);
+
+        // Sanity checks.
+        if (!isExtensionMap1Valid && !isExtensionMap2Valid) return new Dictionary<string, ISet<string>>();
+        if (!isExtensionMap1Valid) return extensionMap2;
+        if (!isExtensionMap2Valid) return extensionMap1;
+
+        // We don't use an additional map for merging, instead we use the first map as a
+        // destination for the merging operation.
+        foreach (var kvp in extensionMap2)
+        {
+            // If the "source" set is empty there's no reason to continue.
```

Comment:
Says that there is no reason to continue, then it continues.

Reply:
The logic below would be better in it's own function that merges sets.

Reply:
Added more details to the comment. What I meant by "there's no reason to continue" is "there's no reason to continue merging for the current key, so we do a `continue` to skip to the next key" :D

Reply:
FIxed the second part, i.e. moved set merging into separate method.